### DISCUSSION TITLE
Bitcoin-Powered Agent for NFT Creation and Auction System

### DISCUSSION THREAD

**Discussion by keyuno** [CounterpartyXCP/CIPs#119] [CIPs]
**Introduction:**

The project consists of three key elements:

1. A Bitcoin-powered agent assigned with the task of generating digital objects (NFTs).

2. An operational live auction system built on the principle of trust minimization.

3. Implementation of atomic swaps to facilitate the completion of transactions.

This project takes the initial steps toward a future in which personal AI agents play a central role in shaping decentralized ecosystems and generating economic value while upholding 'freedom first' as a foundational design principle.

**Bitcoin-powered agent:**

The personal Bitcoin agent exhibits the ability to compose cohesive narratives, interact with other specialized AI systems, and create NFTs.

The process begins with the generation of an address and private keys. Subsequently, the user sends Bitcoin, and upon validation of the transaction by the application, the agent initiates a sequence of actions. In this proof-of-concept, the locally-run language model generates simple fables using randomly selected keywords.

Following this, the agent communicates with another specialized model, instructing it to produce an image that complements the narrative. The process concludes with the broadcast of an IPFS CID (Content Identifier) associated with the generated story.

**Auction System:**

The Counterparty platform primarily revolves around collectibles, and the integration of auctions into the ecosystem is essential. Collectibles and auction houses are inherently interconnected, making the implementation of an auctioning mechanism that isn't dependent on third-party trust crucial.

Auctions can function independently of Bitcoin agents, allowing Counterparty users to auction their assets.

**Conducting an Auction:**

1. The auction system is established through the use of `create_broadcast`.

2. The `fee_fraction` parameter is utilized to determine the XCP participation fee that each bidder must pay to enter the auction.

3. The `value` parameter sets the initial price of the auction, denominated in BTC by default.

5. The duration of the auction is configurable by the seller and is measured in the number of blocks into the future.

6. [Example](https://testnet.xchain.io/tx/34474) of an asset being put up for auction:  

   - A: A601891726002205300, Exp. Block: 2503557

7. Bidders should broadcast the following message (including the `value` in BTC, which must be higher than the starting price):
  

   - B: A601891726002205300
   

**Note**: The current demonstration does not incorporate automated XCP spam prevention. If there is an XCP fee required to participate in the auction, bidders should destroy the specified amount of XCP from the bidding address. When destroying the XCP, they should include a text memo with the name of the auctioned asset.

**Atomic swap:**

Atomic swaps facilitate the transfer of ownership, with the trade being finalized through the exchange of assets between the highest bidder and the seller.

It's important to note that neither party is obligated to follow up with the trade.

**Important Notes:**

Currently, there is a dependency on communicating with the Hugging Face API to generate story illustrations. Ideally, the goal is to establish a decentralized marketplace consisting of various models that a locally-run language model can prompt to generate a wide range of content. This marketplace would operate on the basis of paying for diverse sets of capabilities, utilizing Bitcoin as the payment method to ensure a censorship-resistant process.

The overarching aim is to ensure that AI, progressing towards AGI (Artificial General Intelligence), remains in the hands of individuals and to prevent AI systems from becoming tools of oppression. Decentralization plays a pivotal role as a defensive strategy to accomplish this objective.

Please note that the application currently functions as a proof of concept and will require further development and experimentation to expand its functionality.

**Code Review:**

Presented below is the code for your review:

https://github.com/keyuno/bitcoin-agent

Please note that this code is intended solely for testing and research purposes.

**Comment by davestaxcp** [comment]
I am interested first off how a future User Interface might look for these functions, which sound very interesting and useful.

---

Bitcoin-Powered Agent: Responsible for generating digital objects (NFTs).

- I wonder how IPFS is integrated in this system, especially with wallets like [Looney is working on](https://rarepepewallet.wtf/) and how future UI would show this process. How will it integrate with [CIP-25](https://github.com/CounterpartyXCP/cips/blob/master/cip-0025.md) and how are the  [IPFS integrations](https://github.com/loon3/frog-logger) saved and shared?

---

Operational Live Auction System: Built on the principle of trust minimization.

- I might sound like a broken record, but how would the UI show this? 

- Would the possibility of signing messages to authenticate addresses that might be "allowed" in the auction be a feature?

---

Atomic Swaps: Facilitate the completion of transactions.

**This is the one I am most excited about.**

- Yet again I wonder how much work this would take to integrate into wallets commonly used by the average Counterparty user and how you would show the UI for these functions

- If you were to do an enormous theoretical swap:
1000 separate assets for 10,000 other specific assets in an atomic swap trade, what would the BTC tx fee's look like in that situation?
(I know it's a ridiculous situation, but you never know what uses could arise with this function that extends past DEx functionality)

- How would the tx fee be calculated to include that data in OP_RETURN?

---

I am happy you posted these ideas and I have heard in many different projects chats great interest in these ideas and how they might effect the XCP community.... but of course users just want to test these things with a point and click UI to "see" the functionality at play.,,,, and most users don't comment or make Github accounts.

Even if it is just testnet for now, I think a few users would love to click around an interface to "see" this functionality.

**Comment by keyuno** [comment]
> I am interested first off how a future User Interface might look for these functions, which sound very interesting and useful.

To begin, I envision it as a web browser designed for Web 3.0, similar to the following [example](https://key.studio/#/?id=keyuno-v010). There are numerous potential applications that one can conceptualize.

> Would the possibility of signing messages to authenticate addresses that might be "allowed" in the auction be a feature?

From a user interface perspective, it could be feasible to implement a filter that screens bids coming from addresses associated with particular Identity NFTs.

> I am happy you posted these ideas

It's not just theoretical ideas; there's already a functional testnet node.js application that encompasses all the mentioned features. People simply need to test it out.

![F7F2Rf5WQAA2h67](https://github.com/CounterpartyXCP/cips/assets/37753822/4a068837-9a4c-4eb3-a729-bb09f9028eea)

https://github.com/keyuno/bitcoin-agent
